pkg/monitoring/metrics/virt-handler/migrationdomainstats/queue.go

```diff
@@ -0,0 +1,167 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright the KubeVirt Authors.
+ */
+
+package migrationdomainstats
+
+import (
+	"container/ring"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/tools/cache"
+
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/log"
+
+	domstatsCollector "kubevirt.io/kubevirt/pkg/monitoring/metrics/virt-handler/collector"
+	"kubevirt.io/kubevirt/pkg/monitoring/metrics/virt-handler/domainstats"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/stats"
+)
+
+const (
+	collectionTimeout = 10 * time.Second
+	pollingInterval   = 5 * time.Second
+	bufferSize        = 10
+)
+
+type result struct {
+	vmi       string
+	namespace string
+
+	domainJobInfo stats.DomainJobInfo
+	timestamp     time.Time
+}
+
+type queue struct {
+	vmiStore cache.Store
+
+	vmi     *v1.VirtualMachineInstance
+	results *ring.Ring
+
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	mutex     *sync.Mutex
+	collector domstatsCollector.Collector
+}
+
+func newQueue(vmiStore cache.Store, vmi *v1.VirtualMachineInstance) *queue {
+	return &queue{
+		vmiStore: vmiStore,
+
+		vmi:     vmi,
+		results: ring.New(bufferSize),
+
+		mutex:     &sync.Mutex{},
+		collector: domstatsCollector.NewConcurrentCollector(1),
+	}
+}
+
+func (q *queue) startPolling() {
+	q.ctx = context.Background()
+	q.ctx, q.ctxCancel = context.WithCancel(q.ctx)
+
+	ticker := time.NewTicker(pollingInterval)
+	go func() {
+		for range ticker.C {
+			select {
+			case <-q.ctx.Done():
+				log.Log.Infof("stopping domain stats collection for VMI %s/%s", q.vmi.Namespace, q.vmi.Name)
+				ticker.Stop()
+				return
+			default:
+				log.Log.Infof("collecting domain stats for VMI %s/%s", q.vmi.Namespace, q.vmi.Name)
+				q.collect()
+			}
+		}
+	}()
+}
+
+func (q *queue) collect() {
+	if q.isMigrationFinished() {
+		q.mutex.Lock()
+		defer q.mutex.Unlock()
+
+		q.ctxCancel()
+		return
+	}
+
+	values, err := q.scrapeDomainStats()
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to scrape domain stats for VMI %s/%s", q.vmi.Namespace, q.vmi.Name)
+		return
+	}
+
+	r := result{
+		vmi:       q.vmi.Name,
+		namespace: q.vmi.Namespace,
+
+		domainJobInfo: *values.MigrateDomainJobInfo,
+		timestamp:     time.Now(),
+	}
+
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+	q.results.Value = r
+	q.results = q.results.Next()
```

Comment:
I'm wondering if we should lock before the `q.scrapeDomainStats()`.
Let me explain: 
1. a `metric.Collect()` is executed while we are scraping the domain, it will lock the queue because it is reading results through `q.all()`.
2. the migration finishes at that moment, and the key is deleted from the handler
3.  the queue lock is released so the last results from `queue.collect()` are written
4. these results are never consumed

Maybe it's an edge case, since in the next ticker the `queue.collect()` will notice that the migration is finished and will cancel the ticker, but those results are never picked.
Are we fine with that?
Another (maybe worst) case is if an error occurs while we are reading the informer during point 2, which will cause the vmi to be dropped from the vmiStats loop, while the migration is still in progress.
Is this something that was considered and it is an "acceptable" risk?
Thank you

Reply:
- for the first point, I think it's fine to lose that last bit of data, in terms of user observability about migration, not having that data right before the migration finishes won't affect the outcome significantly (and locking before the scrape can lead to much bigger locking times)

- in terms of removing the vmi, should still be fine, the previous data is still collected, and in the next vmi status update, if the vmi migration is still progressing it will be readded

so IMO we should be fine as we are